core/libdivecomputer.c

```diff
@@ -942,18 +858,57 @@ static void do_save_fingerprint(device_data_t *devdata, const char *tmp, const c
 
 /*
  * Save the fingerprint after a successful download
+ *
+ * NOTE! At this point, we have the final device ID for the divecomputer
+ * we downloaded from. But that 'deviceid' is actually not useful, because
+ * at the point where we want to _load_ this, we only have the libdivecomputer
+ * DC_EVENT_DEVINFO state (devdata->devinfo).
+ *
+ * Now, we do have the devdata->devinfo at save time, but at load time we
+ * need to verify not only that it's the proper fingerprint file: we also
+ * need to check that we actually have the particular dive that was
+ * associated with that fingerprint state.
+ *
+ * That means that the fingerprint save file needs to include not only the
+ * fingerprint data itself, but also enough data to look up a dive unambiguously
+ * when loading the fingerprint. And the fingerprint data needs to be looked
+ * up using the DC_EVENT_DEVINFO data.
+ *
+ * End result:
+ *
+ *   - fingerprint filename depends on the model name and 'devinfo.serial'
+ *     so that we can look it up at DC_EVENT_DEVINFO time before the full
+ *     info has been parsed.
+ *
+ *   - the fingerprint file contains the 'diveid' of the finegrprinted dive,
+ *     whish is just a hash of ther fingerprint itself.
+ *
+ *   - we also save the final 'deviceid' in the fingerprint file, so that
+ *     looking up the dive associated with the fingerprint is possible.
  */
 static void save_fingerprint(device_data_t *devdata)
 {
+	uint32_t model, serial;
 	char *dir, *tmp, *final;
 
-	if (!devdata->fingerprint)
+	// Don't try to save nonexistend fingerprint data
+	if (!devdata->fingerprint || !devdata->fdiveid)
+		return;
+
+	// Don't save fingerprint for no dives downloaded
+	if (!devdata->download_table->nr)
```

Comment:
Not sure, but if we return here, then `dev_data->fingerprint` is never freed..?

Also where are the other device_data_t members freed...?